pam_ldapdb.cpp

```diff
@@ -109,14 +110,29 @@ PAM_EXTERN int pam_sm_authenticate(pam_handle_t* pamh,
     // ensure uri and binddn PAM parameters were specified
     if (arguments.find("uri") == arguments.end() ||
         arguments.find("binddn") == arguments.end()) {
+        pam_syslog(pamh, LOG_NOTICE, "unable to find URI and/or BINDDN");
         return PAM_AUTH_ERR;
     }
 
     std::string binddn = arguments["binddn"];
     replace_all(binddn, "%s", user);
 
+    // ldap_simple_bind_s accepts empty passwords for all users, therefore we
+    // catch and deny them here...
+    if (strlen(pass) == 0) {
+        pam_syslog(pamh, LOG_NOTICE, "ldap authentication failure: "
+                   "empty password for user %s", user);
+        return PAM_AUTH_ERR;
+    }
+
     // check against ldap database
-    return verify(arguments["uri"].c_str(), binddn.c_str(), pass);
+    ret = verify(arguments["uri"].c_str(), binddn.c_str(), pass);
+    if (ret != PAM_SUCCESS) {
+        pam_syslog(pamh, LOG_NOTICE, "ldap authentication failure: "
+                   "user=<%s> pass=<%s> uri=<%s> binddn=<%s>",
```

Comment:
I'm not comfortable with passwords from failed password attempts being written to the NOTICE log. How would you feel about removing `pass=<%s>` here?

Maybe below this call to `pam_syslog` passwords can be included in the DEBUG log:
```c++
pam_syslog(pamh, LOG_NOTICE, "ldap authentication failure: "
                   "user=<%s> uri=<%s> binddn=<%s>",
                   user, arguments["uri"].c_str(), binddn.c_str());
pam_syslog(pamh, LOG_DEBUG, "ldap authentication failure: "
                   "user=<%s> pass=<%s> uri=<%s> binddn=<%s>",
                   user, pass, arguments["uri"].c_str(), binddn.c_str());
```